src/pip/_internal/network/download.py

```diff
@@ -125,12 +133,24 @@ def _get_http_response_filename(resp, link):
     return filename
 
 
-def _http_get_download(session, link):
-    # type: (PipSession, Link) -> Response
+def _http_get_download(session, link, head=False):
```

Comment:
The result of HEAD and GET seem so different that I would wonder if instead we would want to make this method's error handling into a `@contextmanager`. This is also driven by the thought that we might want to extract `link.url.split('#', 1)[0]` into its own named method, because it's not immediately clear what it's doing if you don't know or remember that pypi urls are structured this way.

To avoid breaking other callers, we could keep `_http_get_download()` undisturbed, making use of the new url splitting method and wrapping the `session.get(...)` call in the new contextmanager. That is just what first came to my mind when thinking of how to reuse the error-handling mechanisms in this method -- not a blocking review.

Reply:
I understand the utility of factoring this away into a method though, so `def _http_head(session, link):`, which wraps its `session.head(...)` in the error-handling contextmanager could perhaps be a way to keep the `_FileToDownload` object clean. Just suggestions.

Reply:
> we might want to extract `link.url.split('#', 1)[0]` into its own named method, because it's not immediately clear what it's doing if you don't know or remember that pypi urls are structured this way.

I also think that this is a good idea—the 1/0 numbers looks really cryptic.  I'll add an reference to [PEP 503](https://www.python.org/dev/peps/pep-0503/#specification) to clarify it for future readers.

Regarding splitting a separate method for HEAD, however, I don't *feel* that it would make the callers' code any cleaner, i.e. [this level of abstraction feel just right for me, although it's just personal taste](https://twitter.com/nedbat/status/1299450077967196163).  I'll reword s/head=/just_head=/ to hopefully makes it make more sense.

Reply:
Regarding `head`, it won’t make the caller’s code cleaner, but is still a good idea since having a boolean flag switching between behaviours is a smell in general. it is better to extract the `session.method(...)` call out of this function.

Reply:
@cosmicexplorer and @uranusjr, does 499064d81 make it look better to you?